tests: boardfarm_plugins: add boardfarm tests template

taurha opened the pull request:
Added boardfarm plugin's usage description in README.md.
According to plugin's behavior, it was created example of plugin,
which contains suggested file hierarchy, common necessary files,
with stub test to demonstrate its workflow.

Signed-off-by: Oleksii Ponomarenko <o.ponomarenko@inango-systems.com>

arnout commented on tests/boardfarm_plugins/README.md:
```diff
@@ -0,0 +1,113 @@
+* `name` is any unique name at all
+* `model` is a descriptive model name
+* `conn_cmd` is the command to run to connect to the board
+* `devices` is an array of devices that will be added, must have a WAN device at minimum
```
Is a WAN device really required? Why?

taurha replied:
> Is a WAN device really required? Why?

By the moment, README.md is still in progress and that list of parameters may change when all acceptance criteria will be completed.

arnout commented on tests/boardfarm_plugins/README.md:
```diff
@@ -0,0 +1,113 @@
+As template for the device we can use `<boardfarm-repo-root>/devices/base.py`.
+
+### Test suites
+Custom test suite shoud be presetn in root of boardfarm plugin dir and its name must
```
```suggestion
Custom test suite should be present in root of boardfarm plugin dir and its name must
```

arnout commented on tests/boardfarm_plugins/README.md:
```diff
@@ -0,0 +1,113 @@
+### Boardfarm plugin dir structure:
```
I realize we don't have it in any other markdown file, but we really should have a copyright header:

```suggestion
<!--
SPDX-License-Identifier: BSD-2-Clause-Patent
Copyright (c) 2020 the prplMesh contributors
This code is subject to the terms of the BSD+Patent license.
See LICENSE file for more details.
-->

### Boardfarm plugin dir structure:
```

Note that this would be the first time we use that "the prplMesh contributors" approach. If you feel uncomfortable with it, you can also just use your own name.

taurha replied:
Got it.
"the prplMesh contributors" is OK.

arnout reviewed (commented):
To be honest, I see little added value of a boardfarm_sample_plugin in the repository. Instead I think we should work directly towards starting the prplmesh docker containers from boardfarm, like is done in #1104.

So I'd rework this PR to just add the README file.

arnout commented:
> To be honest, I see little added value of a boardfarm_sample_plugin in the repository. Instead I think we should work directly towards starting the prplmesh docker containers from boardfarm, like is done in #1104.

So, to be clear: it *is* useful to have an "empty" plugin and a simple test that always passes. However, it is not useful to call it `boardfarm_sample_plugin`. Instead, it should already be the prplmesh plugin, but with stub implementations.

To clarify this, I've rephrased the exit criteria of #1080 as follows:

- prplmesh Boardfarm test suite
- prplmesh initial_ap_config test case (but without any test, it just passes)
- Boardfarm device class for prplMesh docker
- Boardfarm test setup config for DUT in docker
- README file with info needed for developers who will migrate `test_flows.py` to Boardfarm

> So I'd rework this PR to just add the README file.

And this I said simple because I think the README is ready and we can merge it already. I prefer to merge things quickly instead of waiting for weeks for a super-PR to be finished.

If you can already reach the rest of the exit criteria in e.g. half a day, then by all means do it in the same PR.

taurha commented:
> I think the README is ready and we can merge it already

It is still in progress as I want to insure that info provided here will be relevant when all other acceptance criteria will be reached.
I'll try to make it done ASAP.

arnout commented:
> > I think the README is ready and we can merge it already
> 
> It is still in progress as I want to insure that info provided here will be relevant when all other acceptance criteria will be reached.
> I'll try to make it done ASAP.

Understood. In that case, don't hurry, and continue as you were planning.

MuteSpirit commented on tests/boardfarm_plugins/README.md:
````diff
@@ -0,0 +1,190 @@
+```
+sudo cp -r boardfarm/html /usr/local/lib/python3.6/dist-packages/boardfarm-1.1.1-py3.6.egg/boardfarm/
+```
+(They were missed in setup script)
````
Maybe include into `boardfarm_hotfix.patch` also changes in `setup.py` script to fix installation `html` files?
They maybe installed as [additional data files](https://docs.python.org/3/distutils/setupscript.html#installing-additional-files)

MuteSpirit commented on tests/boardfarm_plugins/README.md:
````diff
@@ -0,0 +1,190 @@
+
+### Boardfarm plugin dir structure:
+```
+boardfarm_*                 <--- that name template is mandatory for boardfarm plugin
````
I suggest to make comment more strict, i.e.
`"boardfarm_" prefix is mandatory for boardfarm plugin name`

MuteSpirit commented on tests/boardfarm_plugins/run_bf.sh:
```diff
@@ -0,0 +1,15 @@
+#!/bin/bash
+
```
Copyright is missing

MuteSpirit commented on tests/boardfarm_plugins/run_bf.sh:
```diff
@@ -0,0 +1,15 @@
+#!/bin/bash
+
+scriptdir=$(dirname $(readlink -f ${0}))
+bf_plugin_dir=${scriptdir}
```
`bf_plugin_dir` -> `bf_plugins_dir`
Reason: `tests/boardfarm_plugins/` is folder for many plugins

MuteSpirit commented on tests/boardfarm_plugins/run_bf.sh:
```diff
@@ -0,0 +1,15 @@
+#!/bin/bash
+
+scriptdir=$(dirname $(readlink -f ${0}))
```
Please fix shellcheck issue:
```
In test.sh line 2:
scriptdir=$(dirname $(readlink -f ${0}))
                    ^-----------------^ SC2046: Quote this to prevent word splitting.
                                  ^--^ SC2086: Double quote to prevent globbing and word splitting.
```

MuteSpirit commented on tests/boardfarm_plugins/README.md:
```diff
@@ -0,0 +1,190 @@
+            {
+                "name": "repeater1",
+                "type": "prplmesh_docker",
+                "role": "agent",
```
As far as i understand 
1) we are going to use similar test setup config but set another types: `prplmesh_docker` -> `prplwrt`, `radio_dummy` -> `radio_hw` and etc.
2) one container (or one board) may run both controller and agent
See peace of `test_gw_repeater.sh` for run gateway:
```
    [ "$START_GATEWAY" = "true" ] && {
        status "Start GW (Controller + local Agent)"
        "${rootdir}"/tools/docker/run.sh -u "${UNIQUE_ID}" ${VERBOSE_OPT} ${FORCE_OPT} ${GW_EXTRA_OPT} \
            start-controller-agent -d -n "${GW_NAME}" -m 00:11:22:33 -- "$@"
    }
```

So i think that ability to use only single **role** for device is not correct and it's needed to have ability to say that in docker (or board) we want to run both controller and agent

taurha replied:
> 1. we are going to use similar test setup config but set another types: `prplmesh_docker` -> `prplwrt`, `radio_dummy` -> `radio_hw` and etc.

Yeap, right.

> So i think that ability to use only single **role** for device is not correct and it's needed to have ability to say that in docker (or board) we want to run both controller and agent

I'll fix that and add `"role": "both"` for cases when we need both roles at single device.

taurha replied:
@MuteSpirit , I've checked `test_gw_repeater.sh` code - option `start-controller-agent` is the only available to start controller.
As `test_flows.py` apparently using only these two options: `start-controller-agent` and `start-agent` - I think, we should maintain only these ones so far..

arnout replied:
I think we more or less decided that we would use repeater1 as the DUT and always use a docker for the controller - at least to start with. The reasoning behind this is that the controller functionality anyway gets tested even if it runs in docker, and it shouldn't make a difference whether it's in docker or on an actual device.

MuteSpirit commented on tests/boardfarm_plugins/README.md:
```diff
@@ -0,0 +1,190 @@
+* `devices` is a list of devices that will be added, must have at least one device.
+Mandatory fields for child devices(those are listed in **devices** section of parent one):
+* `name` is any unique name at all
+* `board_type` is a name of existing device class
```
For devices it's needed to provide `type` instead of `board_type`

MuteSpirit commented on tests/boardfarm_plugins/README.md:
```diff
@@ -0,0 +1,190 @@
+        "role": "controller",
+        "delay": 7,
+        "conn_cmd": "",
+        "cleanup_cmd": "docker container rm -f gateway",
```
You incapsulates in class `PrplMeshDocker` that controller/agent will run inside docker container.
So i suggest to execute command for stop docker container implicitly in class `PrplMeshDocker` and avoid mention `docker container rm` in test setup config file.

There will be one more advantage in this solution - we won't duplicate container name in fields `name` and `cleanup_cmd`

MuteSpirit commented on tests/boardfarm_plugins/README.md:
```diff
@@ -0,0 +1,190 @@
+As template for the device we can use `<boardfarm-repo-root>/devices/base.py`.
+
+### Test suites
+Custom test suite shoud be presetn in root of boardfarm plugin dir and its name must
```
Fix the typo: `presetn` -> `present`

MuteSpirit commented on tests/boardfarm_plugins/README.md:
```diff
@@ -0,0 +1,190 @@
+### Test suites
+Custom test suite shoud be presetn in root of boardfarm plugin dir and its name must
+match `testsuites*.cfg`.
+Suite names are specified in brackets [], test names must match tests found
```
I suggest to edit phrase next way:
`test names must match test classes names defined in the files at the "boardfarm_*/tests/" directory`

MuteSpirit commented on tests/boardfarm_plugins/boardfarm_sample_plugin/devices/prplmesh_docker.py:
```diff
@@ -0,0 +1,126 @@
+#!/usr/bin/env python3
+
```
Copyright missing

MuteSpirit commented on tests/boardfarm_plugins/boardfarm_sample_plugin/devices/prplmesh_docker.py:
```diff
@@ -0,0 +1,126 @@
+
+        if self.role == "controller":
+            # Spawn dockerized controller
+            self._run_shell_cmd(docker_run_cmd + " start-controller-agent")
```
Command `start-controller-agent` starts controller + local agent instead of alone controller

MuteSpirit commented on tests/boardfarm_plugins/run_bf.sh:
```diff
@@ -0,0 +1,15 @@
+fi
+export PYTHONPATH
+
+exec bft -c boardfarm_sample_plugin/prplmesh_config.json -n prplmesh_dummy -x sample-suite
```
Does i correctly understand that if you pass args (`-c`, `-n`, `-x`) to `bft` here then you are not planning that `run_bf.sh` will be used as Boardfarm run script?

taurha replied:
> Does i correctly understand that if you pass args (`-c`, `-n`, `-x`) to `bft` here then you are not planning that `run_bf.sh` will be used as Boardfarm run script?

Script's current version is supposed to be used only for demonstration of sample plugin.
When in #1071 there will be added operational testsuite where I'll provide a bit more advanced launcher script.

taurha commented on tests/boardfarm_plugins/boardfarm_sample_plugin/devices/prplmesh_docker.py:
```diff
@@ -56,11 +62,16 @@ def __init__(self, *args, **kwargs):
+            # Spawn dockerized controller
+            self._run_shell_cmd(docker_run_cmd + " start-controller-agent")
+            time.sleep(config.get("delay", 5))
+            self.al_entity = ALEntityDocker(self.name, is_controller=True)
```
These changes were made by mistake, removed them in next fixup commit.

tomereli commented on .gitignore:
```diff
@@ -7,6 +7,7 @@ _GO_KW
 .kwps
 kw_reports
 __pycache__/
+*.pyc
```
This should be in a separate commit.

arnout replied:
Actually, this shouldn't even be there... This would only be needed for python2 because python3 compiled files are under `__pycache__`.

tomereli commented on tests/boardfarm_plugins/README.md:
```diff
@@ -0,0 +1,190 @@
+
+### Boardfarm installation
+
+1. To obtain boardfarm sources alongside prplMesh ones run:
```
I suggest to also add a Prerequisites section which will explain that google repo needs to be downloaded - you can refer to its README.md - https://gerrit.googlesource.com/git-repo/+/refs/heads/master/README.md

tomereli commented on tests/boardfarm_plugins/README.md:
````diff
@@ -0,0 +1,190 @@
+
+1. To obtain boardfarm sources alongside prplMesh ones run:
+```
+repo init -u https://github.com/prplfoundation/prplMesh-manifest.git
````
I actually think that we should have a single manifest in prpl and use different branches for different projects where needed, so I would change that to just 'manifest.git' (nitpick)

arnout replied:
Renaming an existing repository is not something that is advisable unless there's a very good reason for it.

tomereli replied:
I created this repository and it is not used anywhere yet, so if we ever want to rename it, now's the time. Its not terribly important but normally the manifest repository is just called manifest, don't know why I chose `prplMesh-manifest` :/

tomereli commented on tests/boardfarm_plugins/README.md:
````diff
@@ -0,0 +1,190 @@
+repo sync
+```
+
+2. Apply provided patch(*boardfarm_hotfix.patch*) to fix boardfarm issue with log writing:
````
In this case I suggest we fork boardfarm and fix it in our fork. Users should not be applying patches.

arnout replied:
:+1: 
The original assumption was that we would be able to use boardfarm as is. If not, we need to fork it.

That also allows us to create an upstream PR.

taurha replied:
Added PR to boardfarm fork with changes proposed in patch:
https://github.com/prplfoundation/boardfarm/pull/8

tomereli commented on tests/boardfarm_plugins/README.md:
````diff
@@ -0,0 +1,190 @@
+```
+
+### Devices
+There're bunch of devices already exist in `<boardfarm-repo-root>/boardfarm/devices`,
````
```suggestion
There are several devices already existing in `<boardfarm-repo-root>/boardfarm/devices`,
```

tomereli commented on tests/boardfarm_plugins/README.md:
```diff
@@ -0,0 +1,190 @@
+class SampleTest(BftBaseTest):
+    def runTest(self):
+        '''
+        Main test logic must be implemented here.
```
Is this according to our python coding style?

taurha replied:
> Is this according to our python coding style?

Actually no, but such naming style is the part of boardfarm plugin requirements.
(i.e. it is mandatory)

tomereli commented on tests/boardfarm_plugins/boardfarm_sample_plugin/devices/prplmesh_docker.py:
```diff
@@ -0,0 +1,126 @@
+        if self.role == "controller":
+            # Spawn dockerized controller
+            self._run_shell_cmd(docker_run_cmd + " start-controller-agent")
+            time.sleep(config.get("delay", 5))
```
Delay should be an optional parameter. Why always delay 5 seconds?

taurha replied:
> Delay should be an optional parameter. Why always delay 5 seconds?

We can rewrite that value using setup description in `.json` config.
I saw that delay hardcoded in `tests/environment.py` (7sec), so I've thought that it is reasonable and added it here(slightly decreased).

tomereli replied:
Yes, I think that we should use the values from the config file.

tomereli commented on tests/boardfarm_plugins/boardfarm_sample_plugin/devices/prplmesh_docker.py:
```diff
@@ -0,0 +1,126 @@
+        if self.cleanup_cmd:
+            self._run_shell_cmd(self.cleanup_cmd)
+
+    def _get_rand_mac(self, oct_amount: int):
```
IMO this should go to a different utils (static) class

taurha replied:
> IMO this should go to a different utils (static) class

It contains a "workaround"(constant first octet), so I can't state that method for general use.
Perhaps, I'll leave that "workaround" here and move that function to separate utils class.
Is that acceptable?

tomereli replied:
Yes, that's acceptable. Also leaving it as is acceptable, your call :)

tomereli commented on tests/boardfarm_plugins/boardfarm_sample_plugin/devices/prplmesh_docker.py:
```diff
@@ -0,0 +1,126 @@
+            mac += ":{:02x}".format(random.randint(0, 255))
+        return mac
+
+    def _run_shell_cmd(self, cmd: str = ""):
```
This should also be in a common class, or in the base class.

taurha replied:
> This should also be in a common class, or in the base class.

Added `PrplMeshBase` parent class, added common methods to it.
(`PrplMeshDocker` inherits it)

tomereli commented on tests/boardfarm_plugins/boardfarm_sample_plugin/prplmesh_config.json:
```diff
@@ -0,0 +1,49 @@
+                "name": "repeater1",
+                "type": "prplmesh_docker",
+                "role": "agent",
+                "delay": 7,
```
So we have the delay here - what is it used for?

taurha replied:
> So we have the delay here - what is it used for?

That delay was hardcoded in `tests/environment.py: launch_environment_docker()` and it's using as timeout between starting of dockerized device and creating corresponding ALEntity object.

tomereli commented on tests/boardfarm_plugins/boardfarm_sample_plugin/tests/initial_ap_config.py:
```diff
@@ -0,0 +1,41 @@
+
+import boardfarm
+
+from boardfarm.tests import bft_base_test
```
All this file changes should be in a separate commit (nitpick)

taurha replied:
> All this file changes should be in a separate commit (nitpick)

I changed commit history according @arnout 's suggestions:
1. add README
2. add plugin but without devices or tests
3. add test + testsuite
4. add base class
5. add docker device
6. move common functions from class methods to separate file
6. add run_bft.sh

tomereli commented on tests/boardfarm_plugins/boardfarm_sample_plugin/devices/prplmesh_docker.py:
```diff
@@ -57,11 +57,16 @@ def __init__(self, *args, **kwargs):
                         + " --verbose --detach --force --name {}".format(self.name) \
                         + " --mac {}".format(self.mac)
 
-        if self.role == "controller":
```
Controller alone doesn't need to be supported IMO, we don't support running a standalone controller without a local agent in prplmesh anyway.

arnout commented on tests/boardfarm_plugins/README.md:
```diff
@@ -0,0 +1,182 @@
+{
+    "prplmesh_dummy": {
+        "name": "gateway",
+        "board_type": "prplmesh_docker",
```
AFAIU, this is supposed to be the DUT, and the "devices" section below is for additional stuff that is connected to the DUT and interacts with it (although it seems the boundary is a bit vague, it also seems to represent parts of the DUT e.g. wan and lan).

Also AFAIU, we were planning to run only agent on the physical DUT and always use a docker for the controller.

So, I think the DUT would typically not be the gateway with controller role, but rather a repeater with agent role.

However, since this is just an example, it doesn't matter much.

arnout commented on tests/boardfarm_plugins/README.md:
```diff
@@ -0,0 +1,182 @@
+Mandatory fields for child devices(those are listed in **devices** section of parent one):
+* `name` is any unique name at all
+* `type` is a name of existing device class
+When running boardfarm we can specify that custom config with in following manner:
```
(nitpick) this should be a new paragraph.
```suggestion

When running boardfarm we can specify that custom config with in following manner:
```

arnout commented on tests/boardfarm_plugins/README.md:
```diff
@@ -0,0 +1,182 @@
+### Test file contents:
+Each test must be presented as class, for test file template please refer to:
+`boardfarm/tests/bft_base_test.py`
+Above you can see mandatory class members:
```
```suggestion
Below you can see mandatory class members:
```
at least I think that's what you mean?

arnout commented on tests/boardfarm_plugins/README.md:
```diff
@@ -0,0 +1,182 @@
+
+### Custom device description
+If the tested device is abcent in `boardfarm/boardfarm/devices` dir
+we can specify new one by creating `custom_config.json` file sort of:
```
Please refer to CODINGSTYLE.md for the coding style of markdown documents:

> Lines in markdown documents are not wrapped, but every sentence should start on a new line.
> Wrapping in text files is annoying, because adding a single word can cause the entire paragraph to be rewrapped.
> On the other hand, putting everything on one line is not readable either.
> As a compromise, no fixed wrapping is done.
> Instead, each sentence should be started on a new line.
> This way, the diff will only show the sentences that are really affected by the change.
> For very long sentences, subsentences may also be placed on a separate line.

Please apply this to the entire document.

arnout commented on tests/boardfarm_plugins/README.md:
```diff
@@ -0,0 +1,182 @@
+    def runTest(self):
+        '''
+        Main test logic must be implemented here.
+        '''
```
As @tomereli commented:
```suggestion
    def runTest(self):
        '''Main test logic must be implemented here.'''
```

Cfr. [PEP8 documentation strings](https://www.python.org/dev/peps/pep-0008/#documentation-strings)

arnout reviewed (changes requested):
Unfinished, but I'll want to re-review afterwards so already marking as Changes Requested.

arnout commented on tests/boardfarm_plugins/README.md:
```diff
@@ -0,0 +1,182 @@
+        Method invoked if DUT's consoles don't respond,
+        generally considering device as 'dead'.
+        Here we can provide info about possible reasons and suggestions.
+        '''
```
```suggestion
        '''Method invoked if DUT doesn't respond.
        
        If the DUT's consoles don't respond, generally the device is considered as 'dead'.
        Here we can provide info about possible reasons and suggestions.
        '''
```

Also for the other docstrings please follow PEP8 coding style.

arnout commented on tests/boardfarm_plugins/README.md:
```diff
@@ -0,0 +1,182 @@
+        '''
+        Teardown method, optional for boardfarm tests.
+        Its 'strange' declaration comparing to other test class methods
+        caused only to boardfarm workflow
```
What does this mean?

Are you referring to the `@classmethod` bit? There's nothing strange about that, it's a decorator. [Search the web for some explanation](https://www.tutorialsteacher.com/python/classmethod-decorator)

taurha replied:
> What does this mean?
> 
> Are you referring to the `@classmethod` bit? There's nothing strange about that, it's a decorator. [Search the web for some explanation](https://www.tutorialsteacher.com/python/classmethod-decorator)

I considered `teardown_class` function strange because I don't understand why it was implemented as 'static' one.
Anyway, It seems like that comment is pointless and I'll remove it leaving only function description.

arnout commented on tests/boardfarm_plugins/boardfarm_sample_plugin/__init__.py:
```diff
@@ -0,0 +1,6 @@
+SPDX-FileCopyrightText: 2020 the prplMesh contributors (see AUTHORS.md)
+This code is subject to the terms of the BSD+Patent license.
+See LICENSE file for more details.
+"""
```
(nitpick) if you do it like this, it will be the module (or in this case, package) docstring. We generally don't want to use the license header for the docstring; instead, we'd want an actual docstring. Therefore:
```suggestion
# SPDX-License-Identifier: BSD-2-Clause-Patent
# SPDX-FileCopyrightText: 2020 the prplMesh contributors (see AUTHORS.md)
# This code is subject to the terms of the BSD+Patent license.
# See LICENSE file for more details.
```

arnout commented on tests/boardfarm_plugins/boardfarm_hotfix.patch:
```diff
@@ -0,0 +1,34 @@
+-                    "time": float(ts[1:-1]),
++                    "time": 0,
+                     "text": str(text),
+                     "name": name
```
This doesn't seem right - I think it will replace one bug with another one. I guess the problem was that this would do the wrong thing in case `line` didn't contain the ` [<timestamp>]` part. So I guess the correct fix would be:
```
if line.startswith(' ['):
    line = line[1:]
    ts, text = line.split(']', 1)
    timestamp = float(ts[1:-1])
else:
    line = text
    timestamp = 0.0

combined_list.append({
    "time": timestamp,
    "text": str(text),
...
```

taurha replied:
> This doesn't seem right - I think it will replace one bug with another one. I guess the problem was that this would do the wrong thing in case `line` didn't contain the ` [<timestamp>]` part. So I guess the correct fix would be:
> 
> ```
> if line.startswith(' ['):
>     line = line[1:]
>     ts, text = line.split(']', 1)
>     timestamp = float(ts[1:-1])
> else:
>     line = text
>     timestamp = 0.0
> 
> combined_list.append({
>     "time": timestamp,
>     "text": str(text),
> ...
> ```

Added that solution to PR:
https://github.com/prplfoundation/boardfarm/pull/8

arnout commented on tests/boardfarm_plugins/boardfarm_sample_plugin/devices/__init__.py:
```diff
@@ -0,0 +1,6 @@
+SPDX-FileCopyrightText: 2020 the prplMesh contributors (see AUTHORS.md)
+This code is subject to the terms of the BSD+Patent license.
+See LICENSE file for more details.
+"""
```
Same comment.

arnout commented on tests/boardfarm_plugins/boardfarm_sample_plugin/devices/prplmesh_docker.py:
```diff
@@ -0,0 +1,137 @@
+SPDX-FileCopyrightText: 2020 the prplMesh contributors (see AUTHORS.md)
+This code is subject to the terms of the BSD+Patent license.
+See LICENSE file for more details.
+"""
```
Same comment, should be comment rather than docstring.

arnout commented on tests/boardfarm_plugins/README.md:
````diff
@@ -0,0 +1,182 @@
+        '''
+        print('{}: Teardown...'.format(cls.__class__.__name__))
+```
+For operational example please refer to **boardfarm_sample_plugin** code.
````
As a wrote before: I don't think it makes sense to add a `boardfarm_sample_plugin`. Instead, add the `boardfarm_prplmesh` plugin. We start out with the sample code, and replace that with the test_flows code as we continue doing the boardfarm conversion.

Once the boardfarm conversion is done, there is really no reason to keep a sample plugin in our repository. So it's better to work on the *actual* plugin right away (but fill it in with sample code to start with).

arnout commented on tests/boardfarm_plugins/boardfarm_sample_plugin/tests/initial_ap_config.py:
```diff
@@ -0,0 +1,42 @@
+        Also, it can be used to provide info about possible failure reasons.
+        '''
+        print('{}: Skipping...'.format(self.__class__.__name__))
+        raise boardfarm.exceptions.SkipTest(msg)
```
Do we need to implement this method or is it optional?

If it is optional, I would just remove it. It's mentioned in the README, that's enough.

arnout commented on tests/boardfarm_plugins/boardfarm_sample_plugin/tests/initial_ap_config.py:
```diff
@@ -0,0 +1,42 @@
+        According to boardfarm sources, it is considered as deprecated,
+        and it is suggested to leave it as a stub.
+        '''
+        print('{}: Recovering...'.format(self.__class__.__name__))
```
Same here: if there is a default implementation in BftBaseTest, there's no need to implement it here.

arnout commented on tests/boardfarm_plugins/boardfarm_sample_plugin/tests/initial_ap_config.py:
```diff
@@ -0,0 +1,42 @@
+    def recover(self):
+        '''
+        Method invoked if test fails with an unexpected exception type.
+        According to boardfarm sources, it is considered as deprecated,
```
If it is deprecated, what is the "new" way of recovering from a hard failure?

taurha replied:
> If it is deprecated, what is the "new" way of recovering from a hard failure?

There's no "new" way for current version(v1.1.1), warning message states that it will be provided in future versions.

arnout commented on tests/boardfarm_plugins/boardfarm_sample_plugin/tests/initial_ap_config.py:
```diff
@@ -0,0 +1,42 @@
+class InitialApConfig(bft_base_test.BftBaseTest):
+    '''
+    PrplMesh sample test case, no actual testing, it just passes
+    '''
```
Docstring coding style:
```suggestion
    '''prplMesh sample test case, no actual testing, it just passes.'''
```

arnout commented on tests/boardfarm_plugins/boardfarm_sample_plugin/testsuites.cfg:
```diff
@@ -0,0 +1,7 @@
+# This code is subject to the terms of the BSD+Patent license.
+# See LICENSE file for more details.
+
+[sample-suite]
```
Can we give a better name? E.g. prplmesh-testflows? Or just testflows?

arnout commented on tests/boardfarm_plugins/boardfarm_sample_plugin/prplmesh_config.json:
```diff
@@ -0,0 +1,49 @@
+        "delay": 7,
+        "conn_cmd": "",
+        "cleanup_cmd": "docker container rm -f gateway",
+        "connection_type": "local_cmd",
```
Please limit the contents of this file to what is actually used/required. I believe that would be just cleanup_cmd and connection_type.

arnout commented on tests/boardfarm_plugins/boardfarm_sample_plugin/prplmesh_config.json:
```diff
@@ -0,0 +1,49 @@
+        "connection_type": "local_cmd",
+        "devices": [
+            {
+                "name": "repeater1",
```
Although the README claims that devices have to be declared, is that really true?

taurha replied:
> Although the README claims that devices have to be declared, is that really true?

With no devices boardfarm will raise exception, so we need specify at least one device.

taurha replied:
> Although the README claims that devices have to be declared, is that really true?

I've rechecked that option - we can specify empty device list `"devices": [],` and that actually works, but `"devices": []` list, even if it's empty, must be present anyway
Added note about that to README.

arnout commented on tests/boardfarm_plugins/run_bf.sh:
```diff
@@ -0,0 +1,19 @@
+
+scriptdir=$(dirname "$(readlink -f "${0}")")
+bf_plugins_dir=${scriptdir}
+bf_dir=$(realpath "${scriptdir}/../../../boardfarm/boardfarm")
```
I would put this script one directory higher, in `tests/run_bf.sh`. It will be the typical entrypoint for the developer who wants to run tests, so each subdirectory is more typing.

arnout commented on tests/boardfarm_plugins/run_bf.sh:
```diff
@@ -0,0 +1,19 @@
+fi
+export PYTHONPATH
+
+exec bft -c "${bf_plugins_dir}"/boardfarm_sample_plugin/prplmesh_config.json -n prplmesh_dummy -x sample-suite
```
Does the boardfarm setup really have to be run? Can't we skip it by using
```suggestion
exec ${bf_dir}/bft -c "${bf_plugins_dir}"/boardfarm_sample_plugin/prplmesh_config.json -n prplmesh_dummy -x sample-suite
```

taurha replied:
> Does the boardfarm setup really have to be run? Can't we skip it by using

We need install boardfarm to run it.
I've spent some time to investigate how we can bypass that limitation - it requires changes of boardfarm sources, which may be very time consuming.

arnout replied:
No we don't.

The only thing that is needed is to install the requirements. So `pip3 install -r requirements.txt` is sufficient. Or alternatively, what I did, install the required packages using the distro package manager.

arnout reviewed (changes requested):
The `__init__.py` in the tests subdirectory is missing the copyright header.

I think it would be better to split this PR into a couple of commits: 
1. README 
2. add plugin but without devices or tests
3. add test + testsuite
4. add device
5. add run_bft.sh

However, I don't think it's worth spending too much time on splitting here. That said, please *do* spend some time on writing a good commit message.

I haven't reviewed the `prplmesh_docker.py` file yet, but I wonder if it really belongs to this PR. Indeed, it doesn't get used in the sample test, so it's dead code for the moment.

Commit:
tests: boardfarm_plugins: add info about boardfarm

Added README.md with info for developers who will migrate
test_flows.py to bardfarm.

Signed-off-by: Oleksii Ponomarenko <o.ponomarenko@inango-systems.com>

Commit:
boardfarm_plugins: boardfarm_prplmesh: add dummy plugin

Added dummy prplMesh plugin for boardfarm according to requirements
mentioned in tests/boardfarm_plugins/README.md.

Signed-off-by: Oleksii Ponomarenko <o.ponomarenko@inango-systems.com>

Commit:
boardfarm_plugins: boardfarm_prplmesh: add dummy testsuite

Added test_flows testsuite for boardfarm plugin.
Added initial_ap_config.py dummy with mandatory test class methods.

Signed-off-by: Oleksii Ponomarenko <o.ponomarenko@inango-systems.com>

Commit:
boardfarm_prplmesh: devices: add dockerized device

Added example of dockerized device class which is suitable for creating
controller and agent instances using existing prplMesh scripts.
Added json config that contains basic device properties and describes
setup with single boardfarm device.

Signed-off-by: Oleksii Ponomarenko <o.ponomarenko@inango-systems.com>

Commit:
boardfarm_prplmesh: devices: add prplMesh base device

Added prplMesh abstract device which contains common methods
required by boardfarm framework.

Signed-off-by: Oleksii Ponomarenko <o.ponomarenko@inango-systems.com>

Commit:
tests: run_bf.sh: add boardfarm startup script

Added startup script for prplMesh plugin.

Signed-off-by: Oleksii Ponomarenko <o.ponomarenko@inango-systems.com>

taurha commented:
Updated prplfoundation/boardfarm fork with latest upstream and added PR with changes previously proposed by boardfarm_hotfix.patch (it is no longer necessary, removed it from current PR):
https://github.com/prplfoundation/boardfarm/pull/8

tomereli commented on tests/boardfarm_plugins/README.md:
```diff
@@ -0,0 +1,174 @@
+### Prerequisites
+
+To obtain prplMesh and boardfarm sources a Google repo tool is used.
+Refer to its README for detailed info:
```
```suggestion
Refer to its [README](https://gerrit.googlesource.com/git-repo/+/refs/heads/master/README.md) for detailed info.
```

tomereli commented on tests/boardfarm_plugins/README.md:
```diff
@@ -0,0 +1,174 @@
+
+To obtain prplMesh and boardfarm sources a Google repo tool is used.
+Refer to its README for detailed info:
+https://gerrit.googlesource.com/git-repo/+/refs/heads/master/README.md
```
```suggestion
```

tomereli commented on tests/boardfarm_plugins/README.md:
````diff
@@ -0,0 +1,174 @@
+```
+repo init -u https://github.com/prplfoundation/prplMesh-manifest.git
+repo sync
+repo forall -p -c 'git checkout $REPO_RREV'
````
This is actually optional, it is perfectly fine to work in detached head (which is what repo is doing by default when its saying "no branch") - bottom line, you should add a comment here saying this is optional (nitpick)

tomereli commented on tests/boardfarm_plugins/README.md:
````diff
@@ -0,0 +1,174 @@
+
+2. Install boardfarm to your host:
+```
+sudo python3 setup.py install
````
1. You need to add a 'cd boardfarm` before this commant
2. Is there a way to install it without sudo?

taurha replied:
> 1. You need to add a 'cd boardfarm` before this commant
> 2. Is there a way to install it without sudo?

1. Indeed
2. No, because we need permissions to write in `/usr/local/lib/python3.6/dist-packages` where boardfarm will be installed.

arnout replied:
Actually, you *can* install without sudo, you just need an additional option `--user`:
```suggestion
python3 setup.py install --user
```

arnout replied:
^^^ You can install without sudo

In addition, I really don't think it is necessary to install it. It is sufficient to install the prerequisites with `pip3 install -r requirements.txt`. This is important because installing it means that every change to boardfarm also requires re-install. Don't tell me it's not possible, I tried it :-). It does require some changes, as indicated elsewhere.

tomereli commented on tests/boardfarm_plugins/boardfarm_prplmesh/tests/initial_ap_config.py:
```diff
@@ -0,0 +1,17 @@
+
+    @classmethod
+    def teardown_class(cls):
+        """Teardown method, optional for boardfarm tests."""
```
Same here - the whole commit looks like it was "cut" in the middle. Something wrong with rebase?

taurha replied:
> Same here - the whole commit looks like it was "cut" in the middle. Something wrong with rebase?

Actually - no.
I've splitted all commits to atomic units as were requested by @arnout.
In that particular case - the dummy testsuite was added, its just passes.

tomereli commented on tests/boardfarm_plugins/boardfarm_prplmesh/__init__.py:
```diff
@@ -0,0 +1,4 @@
+# SPDX-License-Identifier: BSD-2-Clause-Patent
```
Are these empty intentionally?

taurha replied:
> Are these empty intentionally?

Yes, all `__init__.py` were added for only purpose: mark any `*.py` file at same directory level as python module.

tomereli commented on tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_docker.py:
```diff
@@ -5,12 +5,12 @@
 
 from environment import ALEntityDocker
 from .prplmesh_base import PrplMeshBase
+from utils import get_rand_mac
```
I suggest to add a class with static functions, then you can just import the class when using one of the functions instead of importing each function on its own.

taurha replied:
> I suggest to add a class with static functions, then you can just import the class when using one of the functions instead of importing each function on its own.

Indeed.
But as mentioned @arnout in his [comment](https://github.com/prplfoundation/prplMesh/pull/1136#discussion_r408968971) to #1136 draft PR - that function will become useless soon, so I'll remove it here at all.

tomereli reviewed (approved):
Still have some comments, but nothing which requires re-review, so approving.

arnout commented on tests/boardfarm_plugins/README.md:
```diff
@@ -0,0 +1,175 @@
+As template for the device we can use `<boardfarm-repo-root>/devices/base.py`.
+
+### Test suites
+Custom test suite should be present in root of boardfarm plugin dir and its name mustmatch `testsuites*.cfg`.
```
```suggestion
Custom test suite should be present in root of boardfarm plugin dir and its name must match `testsuites*.cfg`.
```

arnout commented on tests/boardfarm_plugins/README.md:
````diff
@@ -0,0 +1,175 @@
+        '''Teardown method, optional for boardfarm tests.'''
+        print('{}: Teardown...'.format(cls.__class__.__name__))
+```
+For operational example please refer to **boardfarm_prplmesh** plugin.
````
:+1:

arnout commented on tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_base.py:
```diff
@@ -0,0 +1,49 @@
+class PrplMeshBase(debian.DebianBox):
+    """PrplMesh abstract device."""
+
+    def _run_shell_cmd(self, cmd: str = "", args: list = None, timeout: int = 30):
```
(nitpick) For completeness, you could add **kwargs here, which are passed unchanged to pexpect.run.

arnout commented on tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_base.py:
```diff
@@ -0,0 +1,49 @@
+from boardfarm.devices import debian
+
+
+class PrplMeshBase(debian.DebianBox):
```
What does DebianDevice give that LinuxDevice doesn't?

I tried with LinuxDevice and so far at least, it works.

taurha replied:
> I tried with LinuxDevice and so far at least, it works.

I've checked - no difference when using `LinuxDevice` instead of `DebianBox`.
Replaced `DebianBox` by `LinuxDevice`.

arnout commented on tests/boardfarm_plugins/boardfarm_prplmesh/prplmesh_config.json:
```diff
@@ -0,0 +1,9 @@
+        "name": "dockerized_device",
+        "board_type": "prplmesh_docker",
+        "role": "controller",
+        "conn_cmd": "",
```
I believe this is only needed for DebianBox, not for LinuxDevice...

taurha replied:
> I believe this is only needed for DebianBox, not for LinuxDevice...

Unfortunately - no.
Field `conn_cmd` is mandatory as it has no default value.
There's no dependency on device class, it's only boardfarm parser's issue.
When boardfarm attempts to parse custom JSON config without that field - exception will be thrown.

arnout commented on tests/run_bf.sh:
```diff
@@ -0,0 +1,18 @@
+
+scriptdir=$(dirname "$(readlink -f "${0}")")
+bf_plugins_dir=${scriptdir}/boardfarm_plugins
+bf_dir=$(realpath "${scriptdir}/../../boardfarm/boardfarm")
```
```suggestion
bf_dir=$(realpath "${scriptdir}/../../boardfarm")
```

arnout commented on tests/run_bf.sh:
```diff
@@ -0,0 +1,18 @@
+fi
+export PYTHONPATH
+
+exec bft -c "${bf_plugins_dir}"/boardfarm_prplmesh/prplmesh_config.json -n prplmesh_docker -x test_flows
```
```suggestion
exec ${bf_dir}/bft -c "${bf_plugins_dir}"/boardfarm_prplmesh/prplmesh_config.json -n prplmesh_docker -x test_flows
```

With those two changes, it works for me without calling `setup.py`.

arnout commented:
Sorry, I pushed "Submit" before writing my overall commit.

There's just two changes I'd still like to see:
1. Modify the run_bf script so it uses boardfarm directly without requiring install. Note that it is not strictly necessary to modify README.md - it's still convenient if boardfarm is installed.
2. Check if LinuxDevice works as well as DebianBox, and if not, add an explanation somewhere (e.g. in a comment in PrplMeshBase) what we use from DebianBox.

arnout commented:
Oh, and there's one more thing: you need to add yourself to AUTHORS.md

Commit:
AUTHORS.md: add contributor

Added contributor entry for Oleksii Ponomarenko.

Signed-off-by: Oleksii Ponomarenko <o.ponomarenko@inango-systems.com>

taurha commented:
> 1. Modify the run_bf script so it uses boardfarm directly without requiring install. Note that it is not 
strictly necessary to modify README.md - it's still convenient if boardfarm is installed.

Updated run_bf script, so now it uses boardfarm local sources instead of installed module.

> 2. Check if LinuxDevice works as well as DebianBox, and if not, add an explanation somewhere (e.g. in a comment in PrplMeshBase) what we use from DebianBox.

I checked and there's no difference of using `LinuxDevice` instead of `DebianBox`, so I changed parent of `PrplMeshBase` class to  `LinuxDevice` as it`s higher in class hierarchy.

arnout reviewed (approved):
Let's get this merged!
